### PR TITLE
1101 refactor matcher logic

### PR DESCRIPTION
**Pull request by davidgisbey**
Context
There was duplication between the custom matchers in the request and controller spec files, for provider and courses.

Changes proposed in this pull request
For both course and provider:

Matchers moved into a matcher folder
Matcher refactored
Tests for pagination  refactored to use objects rather than strings

### CODE REVIEW

File: spec/controllers/api/v1/courses_controller_spec.rb [hunk 1]
```diff
@@ -133,7 +96,7 @@ def format_timestamp(timestamp)
       describe 'returned courses in JSON' do
         subject { response.body }
 
-        it { should_not have_courses }
+        it { should have_courses([]) }
```
Comment (timabell) [inline spec/controllers/api/v1/courses_controller_spec.rb]: me no likey

Comment (defong) [reply spec/controllers/api/v1/courses_controller_spec.rb]: should be_empty ?

**Comment by davidgisbey** [comment]
Additional commit added to use changed_at rather than updated_at in the pagination tests. Interestingly if you create an object using a timestamp with strftime on the updated_at or created_at columns it returns objects with microsecond differences. 

However, if you do this with the changed_at column is creates x objects with an identical timestamp. To  get around this we've used a 1/1000th second difference between the objects being created.

File: spec/controllers/api/v1/courses_controller_spec.rb (hunk 1 above)
Comment (misaka) [reply spec/controllers/api/v1/courses_controller_spec.rb]: Why was `should_not have_courses` removed?

File: spec/controllers/api/v1/providers_controller_spec.rb [hunk 2]
```diff
@@ -133,7 +96,7 @@ def format_timestamp(timestamp)
       describe 'returned providers in JSON' do
         subject { response.body }
 
-        it { should_not have_providers }
+        it { should have_providers([]) }
```
Comment (misaka) [inline spec/controllers/api/v1/providers_controller_spec.rb]: Why no `should_not have_providers`?

**Comment by misaka** [comment]
> Additional commit added to use changed_at rather than updated_at in the pagination tests. Interestingly if you create an object using a timestamp with strftime on the updated_at or created_at columns it returns objects with microsecond differences.
>
> However, if you do this with the changed_at column is creates x objects with an identical timestamp. To get around this we've used a 1/1000th second difference between the objects being created.

That's interesting. I'm going to guess that that FactoryBot does special handling of `created_at` and `updated_at`. This would benefit from Tim's idea to make sure `changed_at` is unique with a unique index.

Also, the 1/1000th second bit is also interesting. I guess this is the issue @defong wrestled with for a bit, it may mean that extending the timestamp used in the HTTP header to be 6 digits is overdoing it, and 3 digits is all we can (easily) support.

File: spec/controllers/api/v1/courses_controller_spec.rb (hunk 1 above)
Comment (davidgisbey) [reply spec/controllers/api/v1/courses_controller_spec.rb]: Have rewrote the logic so should_not works 👍

File: spec/controllers/api/v1/providers_controller_spec.rb (hunk 2 above)
Comment (davidgisbey) [reply spec/controllers/api/v1/providers_controller_spec.rb]: Done 👍

File: spec/controllers/api/v1/courses_controller_spec.rb (hunk 1 above)
Comment (defong) [reply spec/controllers/api/v1/courses_controller_spec.rb]: :rofl: should_not work :roll_eyes:

**Comment by davidgisbey** [comment]
[Refactor controller spec matcher logic into spec helper](https://trello.com/c/M5O2d71f/1101-refactor-controller-spec-matcher-logic-into-spec-helper)

File: spec/support/matchers/have_providers.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
+RSpec::Matchers.define :have_providers do |*providers|
+  providers = providers.flatten
```
Comment (heidar) [inline spec/support/matchers/have_providers.rb]: You could use `Array#flatten!` to flatten in place here. Same in the other matcher.

**Pull request merged**